security: Basic PSA certification level 2 support (#9033)
nrfconnect/sdk-nrf: samples/tfm/provisioning_image/README.rst

.. _provisioning_image:

Provisioning image
##################

.. contents::
   :local:
   :depth: 2

> This sample demonstrates how to initialize the provisioning process of a Trusted Firmware M(TF-M) image.
> The provioning process sets the device in the Provisioning Platform Root of Trust(PRoT) Security Lifecycle as defined in the security model of Platform Security Architecture(PSA).
> This sample does not include a TF-M image, it is a Zephyr image which is required to run before the TF-M image in order to allow the provisioning of the TF-M image later.

Reviewer: ```suggestion
This sample, when executed, will provision the device in a manner compatible with Trusted Firmware M (TF-M).
The device will after execution be in the "PRoT Provisioning" Security Lifecycle state as defined in the "Platform Security Model" of the Platform Security Architecture(PSA).
This sample does not include a TF-M image, it is a Zephyr image which is intended to be flashed, run, and then erased again before the TF-M image is flashed.
```

Reviewer: "in order to allow the provisioning of the TF-M image later"

@Vge0rge : Do we do any provisioning in TF-M later?

Reviewer: Answered my own question. We don't do any provisioning in TF-M. So I think it's misleading to say "in order to allow the provisioning of the TF-M image later".

Reviewer 2: Hm, the transitioning between the provisioning state -> secure state is happening in TF-M. So I cannot say that there is no provisioning happening in TF-M. We basically split the provisioning process, as it is stated in the PSA spec in two steps. The first step is this image and the second step is the TF-M image nrf_provisioning.c logic.

Requirements
************

The following development kits are supported:

.. table-from-sample-yaml::

The :ref:`lib_hw_unique_key` and the :ref:`lib_identity_key` libraries are required to generate and store the Master Key Encryption Key (MKEK) and the identity key into KMU.

.. note::

>    Once the required identity key is provisioned on the device, only the code pages should be erased as ERASEALL removes the identity key from the system.

Reviewer 3: is code pages the term here? to me it means something else.

Reviewer: I suppose internal flash would be less ambiguous.

Reviewer: Or, "the UICR should not be erased, as that removes the identity key from the system"

.. _provisioning_image:

Provisioning image
##################

.. contents::
   :local:
   :depth: 2

This sample, when executed, will initialize the provisioning process of the device in a manner compatible with Trusted Firmware M (TF-M).
The device will, after execution, be in the "PRoT Provisioning" Security Lifecycle state as defined in the "Platform Security Model" of the Platform Security Architecture(PSA).
This sample does not include a TF-M image, it is a Zephyr image which is intended to be flashed, run, and then erased again before the TF-M image is flashed.

When built for the nrf5340dk_nrf5340_cpuapp target this image will also include a child image for the network core (nrf5340dk_nrf5340_cpunet target).

> The child image demonstates how to disable the debugging access on the network core by writing to the UICR->APPROTECT register.

Reviewer 3: ```suggestion
The child image demonstrates how to disable the debugging access on the network core by writing to the UICR->APPROTECT register.
```

Requirements
************

The following development kits are supported:

.. table-from-sample-yaml::

> The :ref:`lib_hw_unique_key` and the :ref:`lib_identity_key` libraries are required to generate and store the Master Key Encryption Key (MKEK) and the identity key into KMU.

Reviewer 4: ```suggestion
The sample also requires the following libraries to generate and store the master key encryption key (MKEK) and the identity key into the key management unit (KMU):

 * :ref:`lib_hw_unique_key`
 * :ref:`lib_identity_key`
```

.. note::

>    Once the required identity key is provisioned on the device the UICR should not be erased, for instance by ERASEALL, as that removes the identity key from the system.

Reviewer 4: ```suggestion
   Once the required identity key is provisioned on the device, the UICR should not be erased by ERASEALL for example, as that removes the identity key from the system.
```

Overview
********

> The PSA security model defines the PRoT Security Lifecycle Provisioning as the state where the device platform security parameters are generated.
> This sample performs the transition from the PRoT Security Lifecycle "Device assembly and test" to the PRoT Security Lifecycle "PRoT Provisioning".
> This enables a TF-M image in a later stage to transition to the PRoT Security Lifecycle "Secured".

Reviewer 4: ```suggestion
The PSA security model defines the PRoT security lifecycle states.
This sample performs the transition from the PRoT security lifecycle state **Device Assembly and Test** to the **PRoT Provisioning** state.

PRoT Provisioning is a state where the device platform security parameters are generated.
This enables a TF-M image to transition to the PRoT security lifecycle state **Secured** at a later stage.
```

The sample performs the following operations:

> 1. The device is verified to be in PRoT Security Lifecycle "Device assembly and test".

Reviewer 4: ```suggestion
1. The device is verified to be in the Device Assembly and Test state.
```

> #. The device is transitioned to the PRoT Security Lifecycle "PRoT Provisioning".

Reviewer 4: ```suggestion
#. The device is transitioned to the PRoT Provisioning state.
```

> #. The random hardware unique keys(HUKs) are generated and stored in the Key Management Unit (KMU).

Reviewer 4: ```suggestion
#. Random hardware unique keys (HUKs) are generated and stored in the key management unit (KMU).
```

#. A random identity key of type secp256r1 is generated and stored in the KMU.

> #. The identity key is verified to be stored in KMU.

Reviewer 4: ```suggestion
#. The identity key is verified to be stored in the KMU.
```

> The Hardware Unique Keys(HUKs) include the Master Key Encryption Key (MKEK) and the Master key for Encrypting External storage (MEXT).

Reviewer 4: ```suggestion
The hardware unique keys (HUKs) include the master key encryption key (MKEK) and the master key for encrypting external storage (MEXT).
```

> The identity key is stored in the KMU in encrypted form using the Hardware Unique Key (HUK) Master Key Encryption Key (MKEK).
> The identity key is used as the Initial Attestation Key as described in the PSA security model.

Reviewer 4: ```suggestion
The identity key is stored in the KMU in encrypted form using the master key encryption key (MKEK).
The identity key is used as the Initial Attestation Key (IAK) as described in the PSA security model.
```

.. _provisioning_image:

Provisioning image
##################

.. contents::
   :local:
   :depth: 2

Running the provisioning image sample will initialize the provisioning process of a device in a manner compatible with Trusted Firmware-M (TF-M).
This sample does not include a TF-M image, it is a Zephyr image intended to be flashed, run, and erased before the TF-M image is flashed.

After completion, the device is in the Platform Root-of-Trust (PRoT) security lifecycle state called **PRoT Provisioning**.
For more information about the PRoT security lifecycle, see Arm's Platform Security Model 1.1 defined in the Platform Security Architecture (PSA).
```

> Should we be linking to something as well here?

Reviewer 4: Remove :)
